File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/AddressComponentsTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class AddressComponentsTyped {
    public String zip;
    public String streetName;
    public String city;
    public String preDirection;
    public String suffixDirection;
    public String state;
    public String suffixType;

    public AddressComponentsTyped(String zip, String streetName, String city, String preDirection, String suffixDirection, String state, String suffixType) {
        this.zip = zip;
        this.streetName = streetName;
        this.city = city;
        this.preDirection = preDirection;
        this.suffixDirection = suffixDirection;
        this.state = state;
        this.suffixType = suffixType;
    }

    public AddressComponentsTyped(AddressComponentsString s) {
        this.zip = s.zip;
        this.streetName = s.streetName;
        this.city = s.city;
        this.preDirection = s.preDirection;
        this.suffixDirection = s.suffixDirection;
        this.state = s.state;
        this.suffixType = s.suffixType;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AddressComponentsTyped)) return false;
        AddressComponentsTyped that = (AddressComponentsTyped) o;
        return Objects.equals(zip, that.zip)
            && Objects.equals(streetName, that.streetName)
            && Objects.equals(city, that.city)
            && Objects.equals(preDirection, that.preDirection)
            && Objects.equals(suffixDirection, that.suffixDirection)
            && Objects.equals(state, that.state)
            && Objects.equals(suffixType, that.suffixType);
    }

    @Override
    public int hashCode() {
        return Objects.hash(zip, streetName, city, preDirection, suffixDirection, state, suffixType);
    }

    public AddressComponentsString toAddressComponentsString() {
        return new AddressComponentsString(String.valueOf(zip), String.valueOf(streetName), String.valueOf(city), String.valueOf(preDirection), String.valueOf(suffixDirection), String.valueOf(state), String.valueOf(suffixType));
    }

    public static List<AddressComponentsTyped> fromStringList(List<AddressComponentsString> list) {
        List<AddressComponentsTyped> result = new ArrayList<>();
        for (AddressComponentsString s : list) result.add(new AddressComponentsTyped(s));
        return result;
    }

    public static List<AddressComponentsString> toStringList(List<AddressComponentsTyped> list) {
        List<AddressComponentsString> result = new ArrayList<>();
        for (AddressComponentsTyped t : list) result.add(t.toAddressComponentsString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("zip", zip);
        m.put("streetName", streetName);
        m.put("city", city);
        m.put("preDirection", preDirection);
        m.put("suffixDirection", suffixDirection);
        m.put("state", state);
        m.put("suffixType", suffixType);
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static AddressComponentsTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new AddressComponentsTyped(
            Json.asString(Json.require(m, "zip"), "zip"),
            Json.asString(Json.require(m, "streetName"), "streetName"),
            Json.asString(Json.require(m, "city"), "city"),
            Json.asString(Json.require(m, "preDirection"), "preDirection"),
            Json.asString(Json.require(m, "suffixDirection"), "suffixDirection"),
            Json.asString(Json.require(m, "state"), "state"),
            Json.asString(Json.require(m, "suffixType"), "suffixType"));
    }

    public static AddressComponentsTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<AddressComponentsTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (AddressComponentsTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<AddressComponentsTyped> fromJSONList(String json) {
        List<AddressComponentsTyped> result = new ArrayList<AddressComponentsTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "AddressComponentsTyped")));
        return result;
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/AddressComponentsTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class AddressComponentsTyped
    {
        public string zip;
        public string streetName;
        public string city;
        public string preDirection;
        public string suffixDirection;
        public string state;
        public string suffixType;

        public AddressComponentsTyped(string zip, string streetName, string city, string preDirection, string suffixDirection, string state, string suffixType)
        {
            this.zip = zip;
            this.streetName = streetName;
            this.city = city;
            this.preDirection = preDirection;
            this.suffixDirection = suffixDirection;
            this.state = state;
            this.suffixType = suffixType;
        }

        public AddressComponentsString ToAddressComponentsString()
        {
            return new AddressComponentsString(
                Json.ToText(this.zip),
                Json.ToText(this.streetName),
                Json.ToText(this.city),
                Json.ToText(this.preDirection),
                Json.ToText(this.suffixDirection),
                Json.ToText(this.state),
                Json.ToText(this.suffixType)
            );
        }

        public static List<AddressComponentsString> ToStringList(List<AddressComponentsTyped> list)
        {
            var result = new List<AddressComponentsString>();
            foreach (var t in list) result.Add(t.ToAddressComponentsString());
            return result;
        }

        public static List<AddressComponentsTyped> FromStringList(List<AddressComponentsString> list)
        {
            var result = new List<AddressComponentsTyped>();
            foreach (var s in list) result.Add(s.ToAddressComponentsTyped());
            return result;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteString("zip", this.zip);
            w.WriteString("streetName", this.streetName);
            w.WriteString("city", this.city);
            w.WriteString("preDirection", this.preDirection);
            w.WriteString("suffixDirection", this.suffixDirection);
            w.WriteString("state", this.state);
            w.WriteString("suffixType", this.suffixType);
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static AddressComponentsTyped FromJsonElement(JsonElement m)
        {
            return new AddressComponentsTyped(
                Json.AsString(Json.Require(m, "zip"), "zip"),
                Json.AsString(Json.Require(m, "streetName"), "streetName"),
                Json.AsString(Json.Require(m, "city"), "city"),
                Json.AsString(Json.Require(m, "preDirection"), "preDirection"),
                Json.AsString(Json.Require(m, "suffixDirection"), "suffixDirection"),
                Json.AsString(Json.Require(m, "state"), "state"),
                Json.AsString(Json.Require(m, "suffixType"), "suffixType")
            );
        }

        public static AddressComponentsTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<AddressComponentsTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<AddressComponentsTyped> FromJSONList(string json)
        {
            var result = new List<AddressComponentsTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "AddressComponentsTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }

        public override string ToString()
        {
            return $"zip={zip}, streetName={streetName}, city={city}, preDirection={preDirection}, suffixDirection={suffixDirection}, state={state}, suffixType={suffixType}";
        }

        public override bool Equals(object? obj)
        {
            if (obj is not AddressComponentsTyped other) return false;
            return object.Equals(this.zip, other.zip)
                && object.Equals(this.streetName, other.streetName)
                && object.Equals(this.city, other.city)
                && object.Equals(this.preDirection, other.preDirection)
                && object.Equals(this.suffixDirection, other.suffixDirection)
                && object.Equals(this.state, other.state)
                && object.Equals(this.suffixType, other.suffixType);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.zip);
            h.Add(this.streetName);
            h.Add(this.city);
            h.Add(this.preDirection);
            h.Add(this.suffixDirection);
            h.Add(this.state);
            h.Add(this.suffixType);
            return h.ToHashCode();
        }
    }
}
